### PR TITLE
Add meta-qualifier @QualifiedLocatioins to restrict usage locations of annotations

### PR DESCRIPTION
**Pull request by topnessman**

### CODE REVIEW

File: framework/src/org/checkerframework/framework/util/BoundTypeUtil.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+        final BoundType boundType;
+        if (typeParamDecl == null) {
+            // This is not only for elements from binaries, but also
+            // when the compilation unit is no longer available.
+            boundType = BoundType.UNKNOWN;
+
+        } else {
+            if (typeParamDecl.getKind() == Tree.Kind.TYPE_PARAMETER) {
+                final TypeParameterTree tptree = (TypeParameterTree) typeParamDecl;
```
Comment (CharlesZ-Chen) [inline framework/src/org/checkerframework/framework/util/BoundTypeUtil.java]: Naming: `tptree` => `tpTree`

Comment (topnessman) [inline framework/src/org/checkerframework/framework/util/BoundTypeUtil.java]: Thanks. I'll do that.

File: framework/src/org/checkerframework/framework/qual/TypeUseLocation.java [hunk 2]
```diff
@@ -81,6 +81,36 @@
...
     IMPLICIT_UPPER_BOUND,
 
+    /** Apply default annotations to unannotated type declarations: {@code @HERE class Demo{}} */
+    TYPE_DECLARATION,
+
+    /**
+     * Represents type argument location in parameterized type {@code List<@TA1 ArrayList<@TA2
+     * String>>}
+     */
+    TYPE_ARGUMENT,
```
Comment (CharlesZ-Chen) [inline framework/src/org/checkerframework/framework/qual/TypeUseLocation.java]: Remove `TYPE_ARGUMENT` and `ARRAY_COMPONENT`.

**Comment by CharlesZ-Chen** [comment]
Update to current upstream, try it on PICO.

**Pull request closed without merging**